// node-opcua-server-configuration/dist/server/install_CertificateAlarm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installCertificateExpirationAlarm = void 0;
/**
 * @module node-opcua-server-configuration-server
 */
const node_opcua_address_space_1 = require("node-opcua-address-space");
const node_opcua_debug_1 = require("node-opcua-debug");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_variant_1 = require("node-opcua-variant");
const debugLog = (0, node_opcua_debug_1.make_debugLog)("ServerConfiguration");
const errorLog = (0, node_opcua_debug_1.make_errorLog)("ServerConfiguration");
const doDebug = (0, node_opcua_debug_1.checkDebugFlag)("ServerConfiguration");
function installCertificateExpirationAlarm(addressSpace) {
    var _a;
    debugLog("installCertificateExpirationAlarm");
    const server = addressSpace.rootFolder.objects.server;
    const namespace = addressSpace.getOwnNamespace();
    const certificateExpirationAlarmType = addressSpace.findEventType("CertificateExpirationAlarmType");
    const options = {
        browseName: "ServerCertificateAlarm",
        conditionSource: null,
        eventSourceOf: server,
        inputNode: new node_opcua_nodeid_1.NodeId(),
        normalState: new node_opcua_nodeid_1.NodeId()
    };
    const data = {};
    const alarm = node_opcua_address_space_1.UACertificateExpirationAlarmImpl.instantiate(namespace, options, data);
    // const alarm = namespace.instantiateOffNormalAlarm({) as UACertificateExpirationAlarm;
    alarm.currentBranch().setRetain(true);
    alarm.activeState.setValue(false);
    alarm.ackedState.setValue(false);
    (_a = alarm.suppressedState) === null || _a === void 0 ? void 0 : _a.setValue(false);
    alarm.certificate.setValueFromSource({ dataType: node_opcua_variant_1.DataType.ByteString, value: null });
    alarm.eventId.setValueFromSource({ dataType: node_opcua_variant_1.DataType.ByteString, value: null });
}
exports.installCertificateExpirationAlarm = installCertificateExpirationAlarm;
